### ISSUE TITLE
[Feat]: for better UX, stream blobs as you fetch when on-demand sync is enabled

### ISSUE THREAD

**Issue by rchincha** [project-zot/zot#1395]
### Is your feature request related to a problem? Please describe.

Currently, we fetch blobs and store locally and then serve. It could cause some clients to error out. So basically, on-demand is mainly a trigger. A better UX for clients would be to stream to them as the blobs are fetched.

### Describe the solution you'd like

For this feature, we may have to enhance or replace the containers/image library entirely and use our own custom library.

You may need a pattern like so (or a combo):
https://pkg.go.dev/io#MultiWriter
https://pkg.go.dev/io#TeeReader

**Comment by rchincha** [comment]
@liangyuanpeng ^ 
https://liangyuanpeng.com/post/new-container-image-registry-of-zot-from-cncf/

**Comment by rchincha** [comment]
A blob that is sync'ed is in three possible states
1. **Found**, so not need to trigger sync on client pulls
2. **NotFound**, so need to trigger sync and add a callback (and store a http.Response ref)
3. **Partial**, sync is still in progress and a client is requesting the blob so it must be added to the list maintained in 2.

Keep in mind locking and memory usage (channels, buffers, etc cost us memory)

**Comment by rchincha** [comment]
https://github.com/project-zot/zot/compare/main...rchincha:zot:issue-1395
^ just some quick pseudo-code

**Comment by peusebiu** [comment]
Hello,

I encountered some issues with fsnotify package and with containers/image(used by sync) package while working on this:
so we create a temp directory (uuid.New()) where containers/image will copy image before commiting it to zot's imagestore.

1st issue with fsnotify is that it doesn't support watching dirs recursively, trying to overcome this by adding subdirs to watchlist whenever a create event comes in, will result in races (a create event could be triggered before adding a subdir to watchlist)

2nd issue with container/image is that they first write to a temp blob file in the form of "oci-put-blob3507476942" before renaming it to the actual blob path.

so I can not identify blobs being written by digest.

basically in sync: we write a blob in a temp file, then move it to a temp directory, then move it to imagestore, trying to keep all these actions synchronized would be really hard.

**Comment by elee1766** [comment]
see: https://pkg.go.dev/github.com/mailgun/multibuf?utm_source=godoc#WriterOnce

**Comment by rchincha** [comment]
For sync feature, add a "stream: true" to enable this ux.

**Comment by rchincha** [comment]
There are pros and cons.

Pros: UX is better

Cons: blob validation etc now falls to the client

**Comment by sizowie** [comment]
+1 for a streaming option

**Comment by sizowie** [comment]
Not stale

**Comment by Agalin** [comment]
It's not even the slow pull start that's the worst part. If zot just kept pushing _anything_ (new lines?) through the connection to keep it open it would already be a much better experience as we wouldn't be hit with HTTP 504 from gateways. Note that [distribution](https://github.com/distribution/distribution/blob/main/registry/proxy/proxyblobstore.go#L82) handles such cases properly and from [quick look at the implementation](https://github.com/distribution/distribution/blob/main/registry/proxy/proxyblobstore.go#L82) it seems to work like this (assuming an image not in the cache):

* Manifest is downloaded then returned to the client. Layers are ignored at this point.
* On layer request:
  * if layer is being downloaded in parallel, the download stream is copied and sent directly to client,
  * if layer hasn't been requested in parallel, its download is started, multiwriter is used to stream to users and local storage at the same time.

Zot downloading blobs in the background during manifest retrieval doesn't really change this logic as any blob can be requested anyway.

**Comment by benoittgt** [comment]
Still important feature.

**Comment by sizowie** [comment]
Reopen pls

**Comment by zf-sebastian-grewe** [comment]
Yes please, we are running into this while testing Zot and this should be considered as a potential fix. Having all clients fail a first non-cache pull isn't great.
